ap51-flash: add support for Alfa AP121F router

ordex opened the pull request:
This PR adds support for the AP121F platform from Alpha. It works by using the netconsole mechanism to send commands to the router.

The only open issue at the moment is that the final "reset" command is not received by the router console (even though it is captured by wireshark). For this reason the final message printed to screen suggests to reboot the device manually.

ecsv commented on router_netconsole.c:
```diff
@@ -29,6 +29,12 @@
+static const unsigned int my_ip = 3232235778UL;  /* 192.168.1.2 */
+
+static const uint8_t zero_mac[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+static const uint8_t ap121f_mac[] = { 0x00, 0x03, 0x7f, 0x09, 0x0b, 0xad };
```
This is not the ap121f_mac but an atheros fallback mac in their u-boot. It is used whenever the actual mac address is not properly loaded by the HW manufacturer in the board specific code.

ecsv commented on router_netconsole.c:
```diff
@@ -40,6 +46,65 @@ struct netconsole_priv {
+}
+
+const struct router_type ap121f = {
+	.desc = "Alpha AP121F",
```
Told you before but you still didn't change the company name

ordex replied:
ops you are right. I am a hard nut to crack. I will fix that.

ecsv commented on router_netconsole.c:
```diff
@@ -29,6 +29,12 @@
+static const unsigned int my_ip = 3232235778UL;  /* 192.168.1.2 */
+
+static const uint8_t zero_mac[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+static const uint8_t ap121f_mac[] = { 0x00, 0x03, 0x7f, 0x09, 0x0b, 0xad };
```
This is not the ap121f_mac but an atheros fallback mac in their u-boot. It is used whenever the actual mac address is not properly loaded by the HW manufacturer in the board specific code. Maybe you should rename this to reflect that.

ordex replied:
interesting...does it mean the manufacturer should make sure the MAC address is written at the proper position so that uboot can fetch it?

Renaming might make sense, but this sounds like other random (different) routers may have the same MAC in the future due to the same problem, right? This would fool the router detection logic.

ecsv replied:
This depends on their initialization code. But yes, they should make sure that they fetch the correct mac address from the flash.

And yes, other routers tend to have the same mac address (due to this problem). If you have some GL-iNet devices (AR750 in my example) then you get the same mac address.

ecsv commented on router_netconsole.c:
```diff
@@ -91,7 +156,7 @@ void handle_netconsole_packet(const char *packet_buff, int packet_buff_len,
 		}
 
-		fprintf(stderr, "[%02x:%02x:%02x:%02x:%02x:%02x]: %s router: rebooting.\n",
+		fprintf(stderr, "[%02x:%02x:%02x:%02x:%02x:%02x]: %s router: rebooting (manual reboot might be required).\n",
```
Did you try to send the "reset" multiple times? Maybe even with some sleeps between the sends (just for testing purposes - I don't really want to have sleeps in the code)

ordex replied:
no I did not. but before sending the reset I ensure we have received the command prompt ("u-boot>"), like before sending the flashing command.

However I will try that just for testing and see what happens.

ecsv replied:
Another idea: you can concat multiple commands with some of the u-boot shells. At least with hush you could do something like:

    run fw_upg;reset\n

Just as a different idea which you could test

ordex replied:
I like this much better! ap51-flash should still be able to understand when the flashing is finished by parsing the output. will give it a try soon

ordex replied:
it worked! awesome :) I'll add a patch for that!

ordex commented:
the pushed branch addresses all the concerns except the one about the MAC. I don't think we can do anything other than reporting that back to the manufacturer.

ecsv commented:
documentation is still missing

ecsv commented on docs/supported-devices/index.rst:
```diff
@@ -16,6 +16,9 @@ quite common amongst a variety of devices.
 Tested
 ======
 
+* Alfa
```
Wrong syntax. This will create following list entry:

* Alfa - AP121F

But we want:

* Alfa
  * AP121F

Please make sure to add an empty line after "Alfa" to achive this

ecsv commented:
@lindnermarek you may want to check whether you want the same workaround for this device as for MR500: https://github.com/ap51-flash/ap51-flash/blob/dbe85bc4e749561532f1e8e52b8376ddae639e9e/flash.c#L168

lindnermarek commented:
@ecsv The mac address check + device state check exists to prevent an endless flash >> reboot >> flash >> reboot cycle. Some devices require pressing some buttons to open the device for a new firmware image. As far as I remember, all OpenMesh devices accept firmware images during boot (without the need to press any buttons).

ecsv commented on router_netconsole.c:
```diff
@@ -64,42 +129,20 @@ void handle_netconsole_packet(const char *packet_buff, int packet_buff_len,
 			node->router_type->desc);
 
-		netconsole_reset(node);
+		priv->state = NODE_STATUS_REBOOTED;
 
-		node->status = NODE_STATUS_REBOOTED;
```
There is now #26 for this slightly odd state mixup. At the moment, GCC reports following:

```
    CC router_netconsole.o
router_netconsole.c: In function ‘handle_netconsole_packet’:
router_netconsole.c:128:29: warning: implicit conversion from ‘enum node_status’ to ‘enum netconsole_state’ [-Wenum-conversion]
  128 |                 priv->state = NODE_STATUS_REBOOTED;
      |
```
